### ISSUE TITLE
Support md-raid

### ISSUE THREAD

**Issue by sergelogvinov** [talos-systems/talos#3607]
## Feature Request

There are many dedicated cloud providers offers to us the servers with two or more disks.
Many of them do not have the hardware raid controllers.

It will be very nice to have softraid on the talos platform.

Thank you.

**Comment by Manouchehri** [comment]
Having a RAID1 setup for all drives to boot off of would be great too.

https://outflux.net/blog/archives/2018/04/19/uefi-booting-and-raid1/

**Comment by clementd64** [comment]
> There are many dedicated cloud providers offers to us the servers with two or more disks. Many of them do not have the hardware raid controllers.

Look at Hetzner, OVH, Scaleway. I understand nodes should be disposable, but if we can avoid failure at first it's always better. 

Also some IO intensive workload require local disk. The local volume can be handled with CSI driver, but no RAID for the rootfs increase points of failure.

**Comment by maxpain** [comment]
Any updates on this?

**Comment by MikeTolkachev** [comment]
I would like to use Talos Linux in my home lab, but I can't.
Since reliability is very important when there are not many servers.

**Comment by maxpain** [comment]
@MikeTolkachev Does your motherboard support hardware RAID in bios?

**Comment by bmalynovytch** [comment]
Not stale

**Comment by divStar** [comment]
To be honest, that's what I also would like for my home-lab.

I currently have two issues with Talos (both possibly being solvable with an md-raid):

1. I'd like to install Talos on two drives and be able to boot from the second drive if the first fails (RAID1).
2. I'd like to combine my 4 U.2 NVMe drives into one storage pool, but with some kind of redundancy level (e.g. RAID5 or RAIDZ1/RAIDZ2).

I am a novice when it comes to Talos and I've been digging for information for quite a few days now. It seems like 2 should be possible, but I don't exactly understand how and 1 is not yet possible? Or am I mistaken?

**Comment by cristianrat** [comment]
This is also a major downer for me. I mean, I am planning on 3 worker nodes, but why waste a perfectly good disk. Servers always come with 2 symmetrical ones

**Comment by maxpain** [comment]
> This is also a major downer for me. I mean, I am planning on 3 worker nodes, but why waste a perfectly good disk. Servers always come with 2 symmetrical ones

Maybe NVMe namespaces could be usable in your scenario

**Comment by cristianrat** [comment]
> > This is also a major downer for me. I mean, I am planning on 3 worker nodes, but why waste a perfectly good disk. Servers always come with 2 symmetrical ones
> 
> Maybe NVMe namespaces could be usable in your scenario

haven't used those before
gonna research them
I imagine talos would work with hardware raid?

**Comment by maxpain** [comment]
> > > This is also a major downer for me. I mean, I am planning on 3 worker nodes, but why waste a perfectly good disk. Servers always come with 2 symmetrical ones
> > 
> > 
> > Maybe NVMe namespaces could be usable in your scenario
> 
> haven't used those before gonna research them I imagine talos would work with hardware raid?

I mean, if you have an enterprise/datacenter NVMe disk, you probably can divide it up into 32 NVMe namespaces. It's like partitions, but on the NVMe firmware level.
In case of a 960GB NVMe disk, you could create a small namespace (let's say 60G) for Talos Linux and it will be visible in the OS as the separate block device (`/dev/nvme0n2`), and another namespace with 920GB for other purposes

**Comment by cristianrat** [comment]
> Yes, Talos will work with hardware RAID.
> And software RAID will come.
> The initial rationale for not implementing software RAID was "cattle not
> pets" - treat your nodes as disposable.
> Instead of mirroring, use one disk dedicated to the OS, and the other for
> workload images, etc.
> In the ideal world this is true, but not always feasible in the practical
> world.
> So RAID will come (sometime.)
> […](#)

fair, and I do
but metal (ie: dedicated) nodes are not cattle :)
but yeah, in the end will stick to regular nodes and just have a smart scan run every X days, and email the hoster to replace the disk if failing (while not replicated, so I'd need to reinstall, it's not that complicated)
open to other ideas

**Comment by cristianrat** [comment]
> > > > This is also a major downer for me. I mean, I am planning on 3 worker nodes, but why waste a perfectly good disk. Servers always come with 2 symmetrical ones
> > > 
> > > 
> > > Maybe NVMe namespaces could be usable in your scenario
> > 
> > 
> > haven't used those before gonna research them I imagine talos would work with hardware raid?
> 
> I mean, if you have an enterprise/datacenter NVMe disk, you probably can divide it up into 32 NVMe namespaces. It's like partitions, but on the NVMe firmware level, so more native. In case of a 960GB NVMe disk, you could create a small namespace (let's say 60G) for Talos Linux and it will be visible in the OS as the separate block device (`/dev/nvme0n1`), and another namespace with 900GB for other purposes (`/dev/nvme0n2`)

cheers for the suggestion. I've not used nvme namespaces before, but they look very handy
most servers come with 2nvme, identical. so... what's the point of namespaces?
I could simply have the os on one (yes, lots of empty space) and the other for volumes etc.

**Comment by maxpain** [comment]
>  lots of empty space
Yes, that's the problem. Talos Linux and some containerd cache usually needs about 10-20G. What if I have 7.68TB disk?

It's better to make a small namespace for Talos Linux, and use the remaining space and the second disk for Linstor, for example.

**Comment by cristianrat** [comment]
> > lots of empty space
> 
> Yes, that's the problem. Talos Linux and some containerd cache usually needs about 10-20G. What if I have 7.68TB disks?
> 
> It's better to make a small namespace for Talos Linux, and use the remaining space and the second disk for Linstor, for example.

You're giving me a lot of good ideas :)
Much obliged

**Comment by smira** [comment]
Talos supports shrinking itself (not using the whole disk for itself), so without NVME namespaces it's easily possible to make Talos smaller, and use the remaining space as you want.

https://www.talos.dev/v1.9/talos-guides/configuration/disk-management/#machine-configuration

**Comment by maxpain** [comment]
> My two cents: I have a Hetzner machine that has two 4TiB HDDs and one 1TiB SSD. Currently I have it configured as a software RAID1 over HDDs and a bcache using SSD a cache on top of that. I am considering switching to Talos but being able to reproduce similar configuration (I am ok with wiping the data) is a must.

My recommendation is to use SDS (Software-Defined Storage) solutions instead of RAID. Take a look at Linstor

**Comment by kshpytsya** [comment]
> > My two cents: I have a Hetzner machine that has two 4TiB HDDs and one 1TiB SSD. Currently I have it configured as a software RAID1 over HDDs and a bcache using SSD a cache on top of that. I am considering switching to Talos but being able to reproduce similar configuration (I am ok with wiping the data) is a must.
> 
> My recommendation is to use SDS (Software-Defined Storage) solutions instead of RAID. Take a look at Linstor or Ceph

I am not sure it will allow me to achieve my goal: treating my storage (2xHDD+SSD) as a single fast and reliable block device holding all partitions: OS (whatever it is) and data.

**Comment by pixply** [comment]
Giving this a push. I have 10 Hetzner Dedicated Servers with two 960GB nvme disks and many ETL pods who are benefitting from high iops. I tried for hours to get find a solution with Talos for a RAID0/1. 

- Custom factory Talos image with mdadm
- Hybrid Raid 1+0 with Raid Superblock at the end with different partitions

Nothing worked

**Comment by cristianrat** [comment]
> Giving this a push. I have 10 Hetzner Dedicated Servers with two 960GB nvme disks and many ETL pods who are benefitting from high iops. I tried for hours to get find a solution with Talos for a RAID0/1.
> 
> * Custom factory Talos image with mdadm
> * Hybrid Raid 1+0 with Raid Superblock at the end with different partitions
> 
> Nothing worked

I would say don't waste your time on this further. They'll support it (hopefully) at some point
I was actually starting to look at rke2 and alternatives, as I want the local iops but also some resiliency

**Comment by pixply** [comment]
> I would say don't waste your time on this further. They'll support it (hopefully) at some point I was actually starting to look at rke2 and alternatives, as I want the local iops but also some resiliency

Yeah, but since the original feature request ist from 2021, my hopes are not too high. I also looked at rke2 or just implementing more disks into the servers and running Talos on its own disk.

**Comment by cristianrat** [comment]
> > I would say don't waste your time on this further. They'll support it (hopefully) at some point I was actually starting to look at rke2 and alternatives, as I want the local iops but also some resiliency
> 
> Yeah, but since the original feature request ist from 2021, my hopes are not too high. I also looked at rke2 or just implementing more disks into the servers and running Talos on its own disk.

fair
One of the reasons I'm running on cloud machines, is lack of raid.
You could run it and chance it, if you have 3 + machines, with redundancy, you  _should_ be ok
(famous last words) :)))

**Comment by CrazyPandar** [comment]
I posted a test result in the Software raid discussion thread. Hope this may help.
https://github.com/siderolabs/talos/discussions/8654#discussioncomment-15092437

**Comment by noe-charmet** [comment]
Omni customer here and this is actually a HUGE pain point for us and causing us quite some trouble.
We have bare metal servers with 2x960GB disks and really want to be able to RAID0 them and have our total usable disk space to be 2TB.

**Comment by kvaps** [comment]
I think we don't even need md-raid here, the needed level of redundancy can be achieved by duplicating operations on both drives simultaneously, eg:

If disks persist:

```yaml
machine:
  install:
    disks: ["/dev/sda", "/dev/sdb"]
```

Then we:
- Create partitions, Install bootloader and talos twise on both of them
- Write machine configuration on both disks

Once a new machine config applied:
- Write it twise on two disks


So in case one disk fail, the system can boot with actual configuration from the other disk, automatcially join to the cluster and download actual etcd and images data.

**Comment by rochecompaan** [comment]
@kvaps you still need md-raid for runtime data on persistent volumes.

**Comment by ArjonBu** [comment]
This is a very important and basic feature. Talos is a great OS, and IMO the biggest use-case for it is running it on bare-metal. But having software raid is extremely important. I was researching Talos for our company but lack of software raid is a deal breaker. 

> I think we don't even need md-raid here, the needed level of redundancy can be achieved by duplicating operations on both drives simultaneously, eg:
> 
> If disks persist:
> 
> machine:
>   install:
>     disks: ["/dev/sda", "/dev/sdb"]
> 
> Then we:
> 
>     * Create partitions, Install bootloader and talos twise on both of them
> 
>     * Write machine configuration on both disks
> 
> 
> Once a new machine config applied:
> 
>     * Write it twise on two disks
> 
> 
> So in case one disk fail, the system can boot with actual configuration from the other disk, automatcially join to the cluster and download actual etcd and images data.

Did you test this?

**Comment by steverfrancis** [comment]
https://github.com/siderolabs/talos/issues/11626
